examples/terraform/ec2/vtctld.tf

```diff
@@ -0,0 +1,105 @@
+resource "aws_network_interface" "vtctld" {
+  subnet_id="${data.aws_subnet_ids.all.ids[0]}"
+  security_groups = ["${aws_security_group.default.id}"]  
+}
+
+resource "aws_eip" "vtctld" {
+  vpc=true
+  network_interface="${aws_network_interface.vtctld.id}"
+}
+
+resource "aws_instance" "vtctld" {
+  ami           = "${data.aws_ami.amazon_linux.id}"
+  instance_type = "${var.instance-type}"
+  key_name = "${aws_key_pair.key.key_name}"
+  
+  tags = {
+    Name = "${var.cluster-name}-vtctld"
+    Cluster = "${var.cluster-name}"
+    Component = "vtctld"
+  }
+
+  network_interface {
+    device_index            = 0
+    network_interface_id    = "${aws_network_interface.vtctld.id}"
+  }
+  
+  provisioner "file" {
+    source="vitess_distro.tgz"
+    destination="/tmp/vitess_distro.tgz"
+
+    connection {
+      type = "ssh"
+      user = "ec2-user"
+    }    
+  }
+
+  provisioner "file" {
+    content= <<EOT
+TOPO_IMPLEMENTATION=etcd2
+TOPO_GLOBAL_ROOT=/vitess/${var.cluster-name}
+TOPO_GLOBAL_SERVER_ADDRESS=${aws_eip.vtctld.private_dns}:2379
+
+CELL_ROOT=/vitess/${var.cluster-name}/${local.cell-name}
+CELL_TOPO_SERVER=${aws_eip.vtctld.private_dns}:2379
+EOT
+    destination="/tmp/vtctld-${local.cell-name}.conf"
+
+    connection {
+      type = "ssh"
+      user = "ec2-user"
+    }    
+  }
+
+  provisioner "file" {
+    content= <<EOT
+[mariadb]
+name = MariaDB
+baseurl = http://yum.mariadb.org/10.3/centos7-amd64
+gpgkey=https://yum.mariadb.org/RPM-GPG-KEY-MariaDB
+gpgcheck=1
+EOT
+    destination="/tmp/MariaDB.repo"
+
+    connection {
+      type = "ssh"
+      user = "ec2-user"
+    }
+  }
+
+  provisioner "remote-exec" {
```

Comment:
While I know that this is a convenient way to start with one instance, I'm actually extremely opposed to the use of terraform provisioners.  I'm opposed to the point that I think it's a pretty dangerous antipattern that terraform would remove if they followed up with people on whether their projects were successful.

In part it's because the infrastructure is tightly coupled with your server configuration, which makes updates extremely difficult, but the bigger issue is testability and reproducibility.

How would you write a test for this?  You would have to run the entire terraform deployment, and then find the servers that were created to confirm they were configured correctly, and terraform already makes it hard to deploy "temporary" environments.

What if some configuration is broken?  Well, the redeploy would fail, which would cause the rest of your terraform run to fail.  I was on a project that used this pattern, and some minor bug prevented us from using terraform to fix our monitoring stack, because that bug happened sooner in the run.

I do get that this is just an example, but I don't want to have such an antipattern that people might copy on here, in part because I don't want to have to debug the future problems that come with it if people use this directly...

I would suggest using something like [molecule](https://molecule.readthedocs.io/en/stable/).  It has some great auto generators, and supports testing on docker, so the development of such a script would be pretty fast.  Then you can have a "post provisioning" step so that terraform is nothing more than the provisioner of machines.

Added bonus is that your work will work on bare metal or other infrastructures for free.